designs/isolate-cluster-upgrades.md

```diff
@@ -0,0 +1,221 @@
+
+We need to clearly document this behavior so users can plan accordingly when updating their workload clusters after a management cluster upgrade.
+
+## Alternative solutions
+### Multiple controller copies
+When upgrading the management cluster, we could keep old versions of the cluster controller for workload clusters (by maintaining versioned deployments). As the workload clusters get upgraded and this old versions become unused, we would delete the corresponding deployments. Each controller would only reconcile clusters tied to their own version. This guarantees a management cluster upgrade won't affect workload clusters since they will still be controlled by the existing version of the controller.
+
+#### Pros
+* Allows drift detection
+* No unexpected rollouts.
+* No special handling business logic in the controller, just a "filtered" watch.
+#### Cons
+* More resource consumption
+* Scales poorly with number of supported versions. With the proposed support model, at most only two versions of the controller need to be maintained, but if this were to change, the maintenance complexity will increase linearly.
+* Increases the complexity by requiring a new component to manage the lifecycle of the cluster controller, something like an EKS-A operator.
+* Requires specially logic to manage controllers other than the `Cluster` controller. Right now all controllers are bundled in the same shared controller manager. If we were gonna have multiple active copies of the cluster controller, we would either need to split it in its own deployment or add special logic to the other controllers so they don't conflict with each other (leader election at the controller level, object sharding, etc.).
+
+### Composable `cluster-api` generation
+Instead of the current monolithic (for lack of a better word) approach to generate the CAPI cluster definition, we could move to a layered model, where each EKS-A version adds another layer on top of the previous one.
+
+```go
+type generationLogic func(capiSpec) capiSpec
+type layer struct {
+	version semver.Version
+	logic []generationLogic
+}
+
+var layers = []layer{
+	{
+		version: semver.Version{Major: 0, Minor: 0, Patch: 0},
+		logic: []generationLogic{baseGeneration},
+	},
+	{
+		version: semver.Version{Major: 0, Minor: 16, Patch: 0},
+		logic: []generationLogic{
+			addNewKubeletFlags,
+			addExtraCyphers,
+		},
+	},
+	{
+		version: semver.Version{Major: 0, Minor: 16, Patch: 3},
+		logic: []generationLogic{
+			skipKubeadmValidation,
+		},
+	},
+}
+
+func generateCAPI(clusterVersion semver.Version) capiSpec {
+	spec := capiSpec{}
+
+	for _, layer := range layers {
+		if clusterVersion.LessThan(layer.version) {
+			break
+		}
+
+		for _, generate := range layer.logic {
+			spec = generate(spec)
+		}
+	}
+
+	return spec
+}
+```
+
+#### Pros
+* Allows drift detection.
+* No unexpected rollouts.
+* Potentially infinite extensibility to support a bigger version skew.
+#### Cons
```

Comment:
Overall I agree that we should go with the proposed solution, since it requires least effort and solve the upgrade problem. Although from pure user/product perspective, this layered solution is more flexible regarding the version it can support, plus that it solves the upgrade problem + does not introduce a problem (drift detection).

I think at least the team can discuss about which solution makes more sense: approach w heavy dev cost and maintenance or simpler solution which also introduce a problem that user can be affected.

Reply:
Agreed, we need alignment here.